File: Stripe/STPPaymentHandlerActionParams.m
//
//  STPPaymentHandlerActionParams.m
//  Stripe
//
//  Created by Yuki Tokuhiro on 6/28/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPPaymentHandlerActionParams.h"

#import "STPAPIClient.h"
#import <Stripe/Stripe3DS2.h>
#import "STPPaymentIntent.h"
#import "STPSetupIntent.h"
#import "STPThreeDSCustomizationSettings.h"
#import "STPThreeDSCustomization+Private.h"

@implementation STPPaymentHandlerPaymentIntentActionParams
{
    BOOL _serviceInitialized;
}

@synthesize threeDS2Service = _threeDS2Service;
@synthesize returnURLString = _returnURLString;

- (instancetype)initWithAPIClient:(STPAPIClient *)apiClient
            authenticationContext:(nullable id<STPAuthenticationContext>)authenticationContext
     threeDSCustomizationSettings:(STPThreeDSCustomizationSettings *)threeDSCustomizationSettings
                    paymentIntent:(STPPaymentIntent *)paymentIntent
                        returnURL:(nullable NSString *)returnURLString
                       completion:(STPPaymentHandlerActionPaymentIntentCompletionBlock)completion {
    self = [super init];
    if (self) {
        _apiClient = apiClient;
        _authenticationContext = authenticationContext;
        _threeDSCustomizationSettings = threeDSCustomizationSettings;
        _returnURLString = [returnURLString copy];
        _paymentIntent = paymentIntent;
        _paymentIntentCompletion = [completion copy];
    }
    
    return self;
}

- (nullable STDSThreeDS2Service *)threeDS2Service {
    if (!_serviceInitialized) {
        _serviceInitialized = YES;
        _threeDS2Service = [[STDSThreeDS2Service alloc] init];
        @try {
            STDSConfigParameters *configParams = [[STDSConfigParameters alloc] initWithStandardParameters];
            if (self.paymentIntent && !self.paymentIntent.livemode) {
                [configParams addParameterNamed:@"kInternalStripeTestingConfigParam" withValue:@"Y"];
            }
            [_threeDS2Service initializeWithConfig:configParams
                                            locale:[NSLocale autoupdatingCurrentLocale]
                                        uiSettings:_threeDSCustomizationSettings.uiCustomization.uiCustomization];
        } @catch (NSException *e) {
            _threeDS2Service = nil;
        }
    }
    
    return _threeDS2Service;
}

- (STPIntentAction *)nextAction {
    return self.paymentIntent.nextAction;
}

- (NSString *)intentStripeID {
    return self.paymentIntent.stripeId;
}

- (void)completeWithStatus:(STPPaymentHandlerActionStatus)status error:(NSError *)error {
    if (self.paymentIntent) {
        NSAssert(self.paymentIntentCompletion != nil, @"Shouldn't have a nil completion block at this point.");
        self.paymentIntentCompletion(status, self.paymentIntent, error);
    } else {
        NSAssert(NO, @"Missing paymentIntent!");
    }
}

@end

@interface STPPaymentHandlerSetupIntentActionParams()
@property (nonatomic) BOOL serviceInitialized;
@property (nonatomic, nullable) STDSThreeDS2Service *threeDS2Service;
@property (nonatomic, nullable, strong) id<STPAuthenticationContext> authenticationContext;
@property (nonatomic, strong) STPAPIClient *apiClient;
@property (nonatomic, strong) STPThreeDSCustomizationSettings *threeDSCustomizationSettings;

@end

@implementation STPPaymentHandlerSetupIntentActionParams

@synthesize returnURLString = _returnURLString;

- (instancetype)initWithAPIClient:(STPAPIClient *)apiClient
            authenticationContext:(nullable id<STPAuthenticationContext>)authenticationContext
     threeDSCustomizationSettings:(STPThreeDSCustomizationSettings *)threeDSCustomizationSettings
                      setupIntent:(STPSetupIntent *)setupIntent
                        returnURL:(nullable NSString *)returnURLString
                       completion:(STPPaymentHandlerActionSetupIntentCompletionBlock)completion {
    self = [super init];
    if (self) {
        _apiClient = apiClient;
        _authenticationContext = authenticationContext;
        _threeDSCustomizationSettings = threeDSCustomizationSettings;
        _returnURLString = [returnURLString copy];
        _setupIntent = setupIntent;
        _setupIntentCompletion = [completion copy];
    }
    
    return self;
}

- (nullable STDSThreeDS2Service *)threeDS2Service {
    if (!_serviceInitialized) {
        _serviceInitialized = YES;
        _threeDS2Service = [[STDSThreeDS2Service alloc] init];
        @try {
            STDSConfigParameters *configParams = [[STDSConfigParameters alloc] initWithStandardParameters];
            if (self.setupIntent && !self.setupIntent.livemode) {
                [configParams addParameterNamed:@"kInternalStripeTestingConfigParam" withValue:@"Y"];
            }
            [_threeDS2Service initializeWithConfig:configParams
                                            locale:[NSLocale autoupdatingCurrentLocale]
                                        uiSettings:_threeDSCustomizationSettings.uiCustomization.uiCustomization];
        } @catch (NSException *e) {
            _threeDS2Service = nil;
        }
    }
    
    return _threeDS2Service;
}

- (STPIntentAction *)nextAction {
    return self.setupIntent.nextAction;
}

- (NSString *)intentStripeID {
    return self.setupIntent.stripeID;
}

- (void)completeWithStatus:(STPPaymentHandlerActionStatus)status error:(NSError *)error {
    if (self.setupIntent) {
        NSAssert(self.setupIntentCompletion != nil, @"Shouldn't have a nil completion block at this point.");
        self.setupIntentCompletion(status, self.setupIntent, error);
    } else {
        NSAssert(NO, @"Missing setupIntent!");
    }
}

@end

File: Stripe/STPPaymentHandlerActionParams.swift
//
//  STPPaymentHandlerActionParams.swift
//  Stripe
//
//  Created by Yuki Tokuhiro on 6/28/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import Foundation

#if canImport(Stripe3DS2)
    import Stripe3DS2
#endif

@available(iOSApplicationExtension, unavailable)
@available(macCatalystApplicationExtension, unavailable)
internal protocol STPPaymentHandlerActionParams: AnyObject {
    var threeDS2Service: STDSThreeDS2Service? { get }
    var threeDS2Transaction: STDSTransaction? { get set }
    var authenticationContext: STPAuthenticationContext { get }
    var apiClient: STPAPIClient { get }
    var threeDSCustomizationSettings: STPThreeDSCustomizationSettings { get }
    var returnURLString: String? { get }
    var intentStripeID: String? { get }
    /// Returns the payment or setup intent's next action
    func nextAction() -> STPIntentAction?
    func complete(with status: STPPaymentHandlerActionStatus, error: NSError?)
}

@available(iOSApplicationExtension, unavailable)
@available(macCatalystApplicationExtension, unavailable)
internal class STPPaymentHandlerPaymentIntentActionParams: NSObject, STPPaymentHandlerActionParams {

    private var serviceInitialized = false

    let authenticationContext: STPAuthenticationContext
    let apiClient: STPAPIClient
    let threeDSCustomizationSettings: STPThreeDSCustomizationSettings
    let paymentIntentCompletion: STPPaymentHandlerActionPaymentIntentCompletionBlock
    let returnURLString: String?
    var paymentIntent: STPPaymentIntent?
    var threeDS2Transaction: STDSTransaction?

    var intentStripeID: String? {
        return paymentIntent?.stripeId
    }

    private var _threeDS2Service: STDSThreeDS2Service?

    var threeDS2Service: STDSThreeDS2Service? {
        if !serviceInitialized {
            serviceInitialized = true
            _threeDS2Service = STDSThreeDS2Service()

            STDSSwiftTryCatch.try(
                {
                    let configParams = STDSConfigParameters()
                    if !(self.paymentIntent?.livemode ?? true) {
                        configParams.addParameterNamed(
                            "kInternalStripeTestingConfigParam", withValue: "Y")
                    }
                    self._threeDS2Service?.initialize(
                        withConfig: configParams,
                        locale: Locale.autoupdatingCurrent,
                        uiSettings: self.threeDSCustomizationSettings.uiCustomization
                            .uiCustomization)
                },
                catch: { _ in
                    self._threeDS2Service = nil
                },
                finallyBlock: {
                })
        }

        return _threeDS2Service
    }

    init(
        apiClient: STPAPIClient,
        authenticationContext: STPAuthenticationContext,
        threeDSCustomizationSettings: STPThreeDSCustomizationSettings,
        paymentIntent: STPPaymentIntent,
        returnURL returnURLString: String?,
        completion: @escaping STPPaymentHandlerActionPaymentIntentCompletionBlock
    ) {
        self.apiClient = apiClient
        self.authenticationContext = authenticationContext
        self.threeDSCustomizationSettings = threeDSCustomizationSettings
        self.returnURLString = returnURLString
        self.paymentIntent = paymentIntent
        self.paymentIntentCompletion = completion
        super.init()
    }

    func nextAction() -> STPIntentAction? {
        return paymentIntent?.nextAction
    }

    func complete(with status: STPPaymentHandlerActionStatus, error: NSError?) {
        paymentIntentCompletion(status, paymentIntent, error)
    }
}

@available(iOSApplicationExtension, unavailable)
@available(macCatalystApplicationExtension, unavailable)
internal class STPPaymentHandlerSetupIntentActionParams: NSObject, STPPaymentHandlerActionParams {
    private var serviceInitialized = false

    let authenticationContext: STPAuthenticationContext
    let apiClient: STPAPIClient
    let threeDSCustomizationSettings: STPThreeDSCustomizationSettings
    let setupIntentCompletion: STPPaymentHandlerActionSetupIntentCompletionBlock
    let returnURLString: String?
    var setupIntent: STPSetupIntent?
    var threeDS2Transaction: STDSTransaction?

    var intentStripeID: String? {
        return setupIntent?.stripeID
    }

    private var _threeDS2Service: STDSThreeDS2Service?

    var threeDS2Service: STDSThreeDS2Service? {
        if !serviceInitialized {
            serviceInitialized = true
            _threeDS2Service = STDSThreeDS2Service()

            STDSSwiftTryCatch.try(
                {
                    let configParams = STDSConfigParameters()
                    if !(self.setupIntent?.livemode ?? true) {
                        configParams.addParameterNamed(
                            "kInternalStripeTestingConfigParam", withValue: "Y")
                    }
                    self._threeDS2Service?.initialize(
                        withConfig: configParams,
                        locale: Locale.autoupdatingCurrent,
                        uiSettings: self.threeDSCustomizationSettings.uiCustomization
                            .uiCustomization)
                },
                catch: { _ in
                    self._threeDS2Service = nil
                },
                finallyBlock: {
                })
        }

        return _threeDS2Service
    }

    init(
        apiClient: STPAPIClient,
        authenticationContext: STPAuthenticationContext,
        threeDSCustomizationSettings: STPThreeDSCustomizationSettings,
        setupIntent: STPSetupIntent,
        returnURL returnURLString: String?,
        completion: @escaping STPPaymentHandlerActionSetupIntentCompletionBlock
    ) {
        self.apiClient = apiClient
        self.authenticationContext = authenticationContext
        self.threeDSCustomizationSettings = threeDSCustomizationSettings
        self.returnURLString = returnURLString
        self.setupIntent = setupIntent
        self.setupIntentCompletion = completion
        super.init()
    }

    func nextAction() -> STPIntentAction? {
        return setupIntent?.nextAction
    }

    func complete(with status: STPPaymentHandlerActionStatus, error: NSError?) {
        setupIntentCompletion(status, setupIntent, error)
    }
}
